src/helpers/ring.rs

```diff
@@ -26,183 +27,270 @@ pub enum HelperAddr {
-                        match buf.entry((item.source, item.type_id)) {
-                            Entry::Occupied(_entry) => {
-                                panic!("Duplicated message {item:?}")
-                            }
-                            Entry::Vacant(entry) => entry.insert(item.payload),
-                        };
-                    }
-                }
-            });
+        fn ring_channel(&self, protocol_id: ProtocolId) -> Self::ChannelType {
+            let table = &mut *self.routing_table.lock().unwrap();
 
-            Self {
-                input_queue: tx,
-                left: None,
-                right: None,
-                buf,
-            }
-        }
-
-        fn set_left(&mut self, left: Sender<MessageEnvelope>) {
-            self.left = Some(left);
-        }
+            // Take ownership of the receivers - if they've been created already, move them out
+            // otherwise create them.
+            let rx = match table.active_routes.entry(protocol_id) {
+                Entry::Occupied(mut entry) => entry.get_mut().inbound.take().unwrap(),
+                Entry::Vacant(entry) => entry.insert(Route::new()).inbound.take().unwrap(),
+            };
 
-        fn set_right(&mut self, right: Sender<MessageEnvelope>) {
-            self.right = Some(right);
+            TestChannel {
+                protocol_id,
+                table: Arc::clone(&self.routing_table),
+                senders: HashMap::from([
+                    (HelperAddr::Left, self.left.clone()),
+                    (HelperAddr::Right, self.right.clone()),
+                ]),
+                receivers: rx,
+            }
         }
     }
 
     #[async_trait]
-    impl Ring for TestHelper {
-        async fn send<T: Message>(&self, dest: HelperAddr, msg: T) -> Result<(), Error> {
-            assert!(self.left.is_some());
-            assert!(self.right.is_some());
-
+    impl Ring for TestChannel {
+        async fn send<T: Message>(&mut self, dest: HelperAddr, msg: T) -> Result<(), Error> {
             // inside the envelope we store the sender of the message (i.e. source)
             // but this method accepts the destination. To obtain source from destination
             // we invert it - message send to the left helper is originated from helper on the
             // right side.
-            let (target, source) = match dest {
-                HelperAddr::Left => (self.left.as_ref().unwrap(), HelperAddr::Right),
-                HelperAddr::Right => (self.right.as_ref().unwrap(), HelperAddr::Left),
+            let source = match dest {
+                HelperAddr::Left => HelperAddr::Right,
+                HelperAddr::Right => HelperAddr::Left,
             };
 
             let bytes = serde_json::to_vec(&msg).unwrap().into_boxed_slice();
             let envelope = MessageEnvelope {
-                type_id: TypeId::of::<T>(),
                 source,
```

Comment:
Do you really need to signal the source?  I guess that - for a test - it's OK to have a little redundancy, but won't the receiver know where the channel comes from?  (If it doesn't, maybe that can be the first message off each new channel...)

Reply:
I can probably remove `source` too but that would make helper setup a bit more complicated. Currently there is one loop in the `TestHelper` that receives all messages sent towards that helper and then dispatches them based on the protocol id value encoded inside the message to `TestChannel` that picks the appropriate receiver. While it is possible to teach `TestHelper` to spawn two tasks and listen to multiple receiver ends and apply the same logic receiving a message, having `source` looks a bit easier to understand